File: C-Plus-Plus/stack/nextSmallestToLeft.cpp
/*
Given an array find the next smallest element to the right
ALGORITHM:
*Initialise a stack and array
*Traverse the given array from the left
*If the stack is empty, then push -1 to array
*If the stack is not empty and topmost element is smaller than th element then push topmost element to array
*If not then pop elements from the stack till it is empty or topmost element is smaller
*/

#include<bits/stdc++.h>
using namespace std;

void nextSmallestToLeft (int a[],int n)
{
    stack <int> s;
    vector <int> v;

    for(int i = 0 ; i < n ; i++)
    {
        if (s.empty())
        v.push_back(-1);
        else if (s.empty() != true && s.top() < a[i])
        v.push_back(s.top());
        else if (s.empty() != true && s.top() > a[i])
        {
            while (s.empty() != true && s.top() > a[i])
            {
                s.pop();
            }
            if (s.empty())
            v.push_back (-1);
            else if (s.top() < a[i])
            v.push_back(s.top());            
        }
        s.push(a[i]);
    }

    for(int i = 0 ; i < n ; i++)
    cout << v[i] << " ";
}

//Driver function
int main()
{
    int n;
    cout << "Enter number of elements: ";
    cin >> n;
    int a[n];
    cout << "Enter the elements:\n";
    for(int i = 0 ; i < n ; i++)
    cin >> a[i];
    nextSmallestToLeft(a,n);
}

/*
Time Complexity: O(n^2)

Space Complexity: O(n)

Input:

Enter number of elements: 4
Enter the elements:
1 2 4 5 

Output:

-1 1 2 4
*/

File: Java/cp/nextSmallestToLeft.java
/*
Given an array of numbers find the next smallest element to the left of each element.
ALGORITHM:
*We use the concept of Last In First Out of stacks
*Here we use the Stack class
*/ 

import java.util.Stack;
import java.util.Scanner;

public class NextSmallestToLeft
{
    public static void nextSmallestToLeft(int a[],int n)
    {
        //Initializing in-built stack
        Stack<Integer> st=new Stack<>();
        int v[]=new int[n];
        for(int i=0;i<n;i++)
        {
            if(st.empty())
            v[i]=-1;
            else if(st.empty()!=true && st.top()<a[i])
            v[i]=st.top();
            else if(st.empty()!=true && st.top()>=a[i])
            {
                while(st.empty()!=true && st.top()>=a[i])
                st.pop();

                if(st.empty())
                v[i]=-1;
                else if(st.top()<a[i])
                v[i]=st.top();
            }
            st.push(a[i]);
        }
        for(int i=0;i<n;i++)
        System.out.print(v[i]+" ");
    }
    public static void main(String args[])
    {
        Scanner s=new Scanner(System.in);
        int n;
        System.out.println("Enter the size of the array: ");
        n=s.nextInt();
        int a[]=new int[n];
        System.out.println("Enter the elements: ");
        for(int i=0;i<n;i++)
        a[i]=s.nextInt();
        nextSmallestToRight(a,n);
        s.close();        
    }
}

/*
Time Complexity: O(n)

Space Complexity: O(n)

Input:

Enter the size of the array: 
5
Enter the elements:
4 6 2 9 0

Output:

-1 4 -1 2 -1
*/
